akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedRunning.scala

```diff
@@ -0,0 +1,377 @@
+    }
+  }
+
+  object PersistingEvents extends EventsourcedRunningPhase {
+    def name = "PersistingEvents"
+    final override def onCommand(c: Command): Behavior[Any] = {
+      log.info(s"PERSISTING EVENTS, command, STASH: ${c}")
+      stash(c, same)
+    }
+    final override def onJournalResponse(response: Response): Behavior[Any] = {
+      log.info("RESPONSE == " + response)
+      response match {
+        case WriteMessageSuccess(p, id) ⇒
+          // instanceId mismatch can happen for persistAsync and defer in case of actor restart
+          // while message is in flight, in that case we ignore the call to the handler
+          if (id == writerIdentity.instanceId) {
+            updateLastSequenceNr(p)
+            popApplyHandler(p.payload)
+            onWriteMessageComplete()
+          } else same
+
+        case WriteMessageRejected(p, cause, id) ⇒
+          // instanceId mismatch can happen for persistAsync and defer in case of actor restart
+          // while message is in flight, in that case the handler has already been discarded
+          if (id == writerIdentity.instanceId) {
+            updateLastSequenceNr(p)
+            onWriteMessageComplete() // FIXME no sure if we should call it here
+            onPersistRejected(cause, p.payload, p.sequenceNr)
+          } else same
+
+        case WriteMessageFailure(p, cause, id) ⇒
+          // instanceId mismatch can happen for persistAsync and defer in case of actor restart
+          // while message is in flight, in that case the handler has already been discarded
+          if (id == writerIdentity.instanceId) {
+            onWriteMessageComplete()
+            onPersistFailureThenStop(cause, p.payload, p.sequenceNr)
+          } else same
+
+        case _: LoopMessageSuccess ⇒
+          // ignore, not used in Typed Persistence (needed for persistAsync)
+          same
+
+        case WriteMessagesSuccessful ⇒
+          writeInProgress = false
+          flushJournalBatch()
+
+        case WriteMessagesFailed(_) ⇒
+          writeInProgress = false
+          same // it will be stopped by the first WriteMessageFailure message
+
+      }
+    }
+
+    private def onWriteMessageComplete(): Behavior[Any] = {
+      tryBecomeHandlingCommands()
+      tryUnstash(context, same)
+    }
+
+    private def onPersistRejected(cause: Throwable, event: Any, seqNr: Long): Behavior[Any] = {
+      log.error(
+        cause,
+        "Rejected to persist event type [{}] with sequence number [{}] for persistenceId [{}] due to [{}].",
+        event.getClass.getName, seqNr, persistenceId, cause.getMessage)
+
+      same
+    }
+
+    private def onPersistFailureThenStop(cause: Throwable, event: Any, seqNr: Long): Behavior[Any] = {
+      log.error(cause, "Failed to persist event type [{}] with sequence number [{}] for persistenceId [{}].",
+        event.getClass.getName, seqNr, persistenceId)
+
+      // FIXME should we throw perhaps instead?
```

Comment:
In untyped we unconditionally stop on persist failures, because we can't know if the event was persisted or not so there must be a new recovery. resume would be wrong. restart would typically result in restart loop. Intended to be used with backoff supervision.

In Typed we have restart with backoff so perhaps we should rethink the stopping for Typed. Perhaps we can always add a restart with backoff for PersistentBehavior that handles the persist failure exceptions. 

Created ticket for this: https://github.com/akka/akka/issues/24479
In this PR we can stop.

Reply:
Ok done